File: scripts/compress_images.py
"""
Compress product images to reduce repository size.
Uses temp file to avoid Windows file locking issues.
"""
import os
import shutil
import tempfile
from PIL import Image

TARGET_DIRS = [
    r'public/assets/images/products/wall_scene_summary',
    r'public/assets/images/products/floor_scene_summary',
    r'public/assets/images/products/wall_main_summary',
    r'public/assets/images/products/floor_main_summary',
]

MAX_DIMENSION = 1200
JPEG_QUALITY = 72

total_original = 0
total_saved = 0
processed = 0
skipped = 0

def compress_file(filepath):
    global total_original, total_saved, processed, skipped

    ext = os.path.splitext(filepath)[1].lower()
    if ext not in ('.jpg', '.jpeg', '.png'):
        return

    original_size = os.path.getsize(filepath)
    if original_size < 50 * 1024:
        skipped += 1
        return

    try:
        # Use temp file to avoid Windows file locking
        tmp_fd, tmp_path = tempfile.mkstemp(suffix='.jpg')
        os.close(tmp_fd)

        img = Image.open(filepath)
        w, h = img.size

        # Resize if needed
        if w > MAX_DIMENSION or h > MAX_DIMENSION:
            ratio = min(MAX_DIMENSION / w, MAX_DIMENSION / h, 1.0)
            new_w, new_h = int(w * ratio), int(h * ratio)
            img = img.resize((new_w, new_h), Image.LANCZOS)

        # Convert RGBA/P to RGB for JPEG
        if img.mode in ('RGBA', 'P'):
            img = img.convert('RGB')

        # Save to temp file
        img.save(tmp_path, 'JPEG', quality=JPEG_QUALITY, optimize=True)
        img.close()

        new_size = os.path.getsize(tmp_path)
        if new_size < original_size:
            # Replace original with compressed version
            os.remove(filepath)
            shutil.move(tmp_path, filepath)
            saved = original_size - new_size
            total_original += original_size
            total_saved += saved
            processed += 1
            pct = (1 - new_size / original_size) * 100
            short_name = os.path.basename(filepath)[:40]
            print(f"  OK {short_name:42s} {original_size//1024:>6}KB -> {new_size//1024:>6}KB ({pct:.0f}%)")
        else:
            os.remove(tmp_path)
            skipped += 1

    except Exception as e:
        print(f"  ERR {os.path.basename(filepath)[:40]:42s} {e}")


def main():
    global processed, skipped, total_original, total_saved

    print("=== Image Compression (Python) ===\n")

    for dirpath in TARGET_DIRS:
        if not os.path.isdir(dirpath):
            print(f"SKIP: {dirpath}")
            continue

        files = sorted(f for f in os.listdir(dirpath)
                       if f.lower().endswith(('.jpg', '.jpeg', '.png')))

        print(f"\n[{dirpath}] ({len(files)} images)")

        for fname in files:
            compress_file(os.path.join(dirpath, fname))

        # Show dir size after
        img_total = sum(os.path.getsize(os.path.join(dirpath, f))
                        for f in os.listdir(dirpath)
                        if f.lower().endswith(('.jpg', '.jpeg', '.png', '.webp')))
        print(f"  -> {img_total / (1024**3):.2f} GB")

    # Individual product images
    products_dir = r'public/assets/images/products'
    print(f"\n[Individual products]")
    if os.path.isdir(products_dir):
        prod_dirs = sorted(d for d in os.listdir(products_dir)
                          if os.path.isdir(os.path.join(products_dir, d))
                          and d[:2].lower() in ('ft', 'mt'))
        for dname in prod_dirs:
            dpath = os.path.join(products_dir, dname)
            for fname in os.listdir(dpath):
                compress_file(os.path.join(dpath, fname))

    print(f"\n=== Summary ===")
    print(f"Processed: {processed}")
    print(f"Skipped: {skipped}")
    if total_original > 0:
        print(f"Original: {total_original / (1024**3):.2f} GB")
        print(f"Saved: {total_saved / (1024**3):.2f} GB")
        print(f"Reduction: {total_saved / total_original * 100:.1f}%")
    else:
        print("No files processed")


if __name__ == '__main__':
    main()

File: scripts/compress_images.mjs
import sharp from 'sharp';
import fs from 'fs';
import path from 'path';

// Target directories to compress
const TARGET_DIRS = [
  'public/assets/images/products/墙贴场景汇总',
  'public/assets/images/products/地贴场景图汇总',
  'public/assets/images/products/墙贴主图汇总',
  'public/assets/images/products/地贴主图汇总',
  'public/assets/images/factory',
];

// Max dimension for resizing
const MAX_DIMENSION = 1200;

// JPEG quality
const JPEG_QUALITY = 75;

// WebP quality
const WEBP_QUALITY = 75;

let totalSaved = 0;
let totalOriginal = 0;
let processedCount = 0;
let skippedCount = 0;

async function compressFile(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  if (!['.jpg', '.jpeg', '.png', '.webp'].includes(ext)) return;

  const stat = fs.statSync(filePath);
  const originalSize = stat.size;

  // Skip files under 100KB
  if (originalSize < 100 * 1024) {
    skippedCount++;
    return;
  }

  try {
    let image = sharp(filePath);
    const metadata = await image.metadata();

    let needsResize = false;
    if (metadata.width > MAX_DIMENSION || metadata.height > MAX_DIMENSION) {
      needsResize = true;
      image = image.resize({
        width: Math.min(metadata.width, MAX_DIMENSION),
        height: Math.min(metadata.height, MAX_DIMENSION),
        fit: 'inside',
        withoutEnlargement: true,
      });
    }

    // For JPEG and PNG: compress as JPEG
    if (ext === '.jpg' || ext === '.jpeg') {
      image = image.jpeg({ quality: JPEG_QUALITY, mozjpeg: true });
    } else if (ext === '.png') {
      // Convert PNG to WebP for better compression
      const webpPath = filePath.replace(/\.png$/i, '.webp');
      await image.webp({ quality: WEBP_QUALITY }).toFile(webpPath);
      const newSize = fs.statSync(webpPath).size;
      if (newSize < originalSize) {
        fs.unlinkSync(filePath);
        totalSaved += originalSize - newSize;
        totalOriginal += originalSize;
        processedCount++;
        console.log(`✓ ${path.basename(filePath)} -> ${path.basename(webpPath)} (${formatSize(originalSize)} -> ${formatSize(newSize)})`);
      } else {
        // WebP is not smaller, keep original
        fs.unlinkSync(webpPath);
        skippedCount++;
      }
      return;
    } else if (ext === '.webp') {
      // Re-compress WebP
      image = image.webp({ quality: WEBP_QUALITY });
    }

    const tmpPath = filePath + '.tmp';
    await image.toFile(tmpPath);
    const newSize = fs.statSync(tmpPath).size;

    if (newSize < originalSize) {
      fs.renameSync(tmpPath, filePath);
      totalSaved += originalSize - newSize;
      totalOriginal += originalSize;
      processedCount++;
      console.log(`✓ ${path.relative('public', filePath)} (${formatSize(originalSize)} -> ${formatSize(newSize)})`);
    } else {
      fs.unlinkSync(tmpPath);
      skippedCount++;
    }
  } catch (err) {
    console.error(`✗ Error processing ${filePath}: ${err.message}`);
  }
}

function formatSize(bytes) {
  if (bytes < 1024) return bytes + 'B';
  if (bytes < 1024 * 1024) return (bytes / 1024).toFixed(1) + 'KB';
  return (bytes / (1024 * 1024)).toFixed(1) + 'MB';
}

async function main() {
  console.log('=== Image Compression Script ===\n');

  for (const dir of TARGET_DIRS) {
    const fullDir = path.resolve(dir);
    if (!fs.existsSync(fullDir)) {
      console.log(`Skipping (not found): ${dir}`);
      continue;
    }

    const files = fs.readdirSync(fullDir)
      .filter(f => /\.(jpg|jpeg|png|webp)$/i.test(f))
      .map(f => path.join(fullDir, f));

    console.log(`\n📁 ${dir} (${files.length} images)`);

    for (const file of files) {
      await compressFile(file);
    }
  }

  // Also compress individual product images (not in summary dirs)
  console.log('\n📁 Individual product images...');
  const productDirs = fs.readdirSync('public/assets/images/products')
    .filter(d => /^mt\d+|^ft\d+/.test(d))
    .map(d => path.join('public/assets/images/products', d));

  for (const dir of productDirs) {
    if (!fs.existsSync(dir)) continue;
    const files = fs.readdirSync(dir)
      .filter(f => /\.(jpg|jpeg|png|webp)$/i.test(f))
      .map(f => path.join(dir, f));

    for (const file of files) {
      await compressFile(file);
    }
  }

  console.log('\n=== Summary ===');
  console.log(`Processed: ${processedCount} files`);
  console.log(`Skipped: ${skippedCount} files`);
  console.log(`Total original: ${formatSize(totalOriginal)}`);
  console.log(`Total saved: ${formatSize(totalSaved)}`);
  console.log(`Reduction: ${totalOriginal > 0 ? ((totalSaved / totalOriginal) * 100).toFixed(1) : 0}%`);
}

main().catch(console.error);
